txpool/txpool.go

```diff
@@ -192,576 +200,280 @@ func NewTxPool(
-			lowestNonceTx.Nonce >= a.accountQueue.nextNonce { // The lowest nonce tx is valid
-			// All good
-			break
+		// check if shutdown was called
+		select {
+		case <-p.shutdownCh:
+			return
+		default:
 		}
 
-		// Remove the transaction
-		a.accountQueue.logger.Debug(
-			fmt.Sprintf(
-				"Dropping txn [%s] due to a low nonce [%d < %d]",
-				lowestNonceTx.Hash.String(),
-				lowestNonceTx.Nonce,
-				a.accountQueue.nextNonce,
-			),
-		)
-		a.accountQueue.txs.Pop()
-
-		pruneCallback(lowestNonceTx)
-	}
-}
-
-// GetNonce returns the next nonce for the account
-// -> Returns the value from the TxPool if the account is initialized in-memory
-// -> Returns the value from the world state otherwise
-func (t *TxPool) GetNonce(addr types.Address) uint64 {
-	// Grab the account queue lock
-	wrapper := t.lockAccountQueue(addr, true)
-	defer wrapper.unlock()
-
-	return wrapper.accountQueue.nextNonce
-}
-
-// GetCapacity returns the current number of slots occupied and the max slot limit
-func (t *TxPool) GetCapacity() (uint64, uint64) {
-	return t.gauge.getHeight(), t.gauge.limit
-}
-
-// NumAccountTxs Returns the number of transactions in the account specific queue
-func (t *TxPool) NumAccountTxs(address types.Address) int {
-	wrapper := t.lockAccountQueue(address, false)
-	defer wrapper.unlock()
+		// handle requests
+		for {
+			select {
+			case req := <-p.enqueueReqCh:
+				go p.handleEnqueueRequest(req)
+			case req := <-p.promoteReqCh:
+				go p.handlePromoteRequest(req)
+			}
+		}
+	}()
 
-	return len(wrapper.accountQueue.txs)
+	return nil
 }
 
-func (t *TxPool) AddSigner(s signer) {
-	t.signer = s
+// Stops the pool's main loop.
+func (p *TxPool) Stop() {
+	p.shutdownCh <- struct{}{}
 }
 
-func (t *TxPool) handleGossipTxn(obj interface{}) {
-	if !t.sealing {
-		return
-	}
-
-	raw := obj.(*proto.Txn)
-	txn := new(types.Transaction)
-	if err := txn.UnmarshalRLP(raw.Raw.Value); err != nil {
-		t.logger.Error("failed to decode broadcasted txn", "err", err)
-	} else {
-		if err := t.addImpl(OriginGossip, txn); err != nil {
-			if !errors.Is(err, ErrAlreadyKnown) {
-				t.logger.Error("failed to add broadcasted txn", "err", err)
-			}
-		}
-	}
+// Sets the signer the pool will use
+// to check a transaction's signature.
+func (p *TxPool) AddSigner(s signer) {
+	p.signer = s
 }
 
-// EnableDev enables dev mode for the txpool
-func (t *TxPool) EnableDev() {
-	t.dev = true
+// Enables dev mode so the pool can accept
+// non-encrypted transactions. (used for testing)
+func (p *TxPool) EnableDev() {
+	p.dev = true
 }
 
-// AddTx adds a new transaction to the pool and broadcasts it if networking is enabled
-func (t *TxPool) AddTx(tx *types.Transaction) error {
-	if err := t.addImpl(OriginAddTxn, tx); err != nil {
+// AddTx adds a new transaction to the pool (sent from json-RPC/gRPC endpoints)
+// and broadcasts it to the network (if enabled).
+func (p *TxPool) AddTx(tx *types.Transaction) error {
```

Comment:
What do you think about renaming it to AddLocalTx?

Reply:
I do agree that `addLocalTx` is what the function is doing, but from the perspective packages using this func, the notion of `local` vs `gossip` should not be of relevance. 

I vode on `AddTx` remaining the same, but changing `handleGossipTx` to `addGossipTx` since the pool is the only module that should be aware of `gossiped` transactions.

Reply:
That's a good argument 👍